### PR TITLE
DLCQuest: logic speed up

### PR DESCRIPTION
**Pull request by Berserker66**
## What is this fixing or adding?
Makes DLCQuest logic go vroom

## How was this tested?
Grabbed the base yaml, enabled coinsanity, set it to 1, copied the file to a total of 3. Ran it, took 45 seconds. After applying my shenanigans it now ran in 10 seconds.

@axe-y @agilbert1412

### CODE REVIEW

File: worlds/dlcquest/Items.py [hunk 1]
```diff
@@ -11,6 +11,8 @@
 
 class DLCQuestItem(Item):
     game: str = "DLCQuest"
+    coins: int = 0
```
Comment (agilbert1412) [inline worlds/dlcquest/Items.py:14]: in 2066, I've tried to add consistency in defining the base campaign stuff either "dlcquest" or simply "basic", as opposed to not giving it any name. This is for it to be easier to distinguish "coin" (dlc quest coin) vs "coin" (any coin). I would recommend following this pattern here as well

File: worlds/dlcquest/__init__.py [hunk 2]
```diff
@@ -87,3 +86,17 @@ def fill_slot_data(self):
             "seed": self.random.randrange(99999999)
         })
         return options_dict
+
+    def collect(self, state: CollectionState, item: DLCQuestItem) -> bool:
+        change = super().collect(state, item)
+        if change:
+            state.prog_items["COINS", self.player] += item.coins
```
Comment (agilbert1412) [inline worlds/dlcquest/__init__.py:93]: same goes for this

**Review by agilbert1412 — APPROVED** [review]
I have no clue why this makes the code faster in any way. But I looked over it, and the code itself seems fine.

It sure would be nice if it could be checked against the unit tests from https://github.com/ArchipelagoMW/Archipelago/pull/2066 though

I suspect that the strategy used here might be usable in Stardew as well, so if you could explain it to me a little bit, it might let me make a big improvement there as well

**Comment by BootsinSoots** [comment]
Wish Granted

Stress Test 
Current 0.4.3 Build - 2:45 minute generation time, Playthrough instantaneous
PR #2066 - 2:45 minute generation time, Playthrough _2:30 minutes_
PR #2323 - 0:45 _seconds_ generation time, Playthrough 0:15 seconds

Would you like me to leave this on 2206 as well? I did notice it has merge conflicts still

**Comment by agilbert1412** [comment]
@BootsinSoots 
> Wish Granted.
Hmmm, I don't think you understood my wish. I didn't mean checking performance against 2066, but more like running the tests introduced in 2066 in order to validate correctness of behavior. I believe that pre-2066 DLCQuest has, essentially, zero tests, so it's impossible for me to validate if this PR introduces bugs or not. Although based on the numbers shown in the rest of your comment, I must admit that checking performance against 2066 seemed to have been **also** very valuable...

> Stress Test Current 0.4.3 Build - 2:45 minute generation time, Playthrough instantaneous PR #2066 - 2:45 minute generation time, Playthrough _2:30 minutes_ PR #2323 - 0:45 _seconds_ generation time, Playthrough 0:15 seconds

Wait, so both PRs have worse performance on the playthrough generation step? That is curious.
As I said, I don't understand 2323 completely, so maybe that's a tradeoff Berserker understands and decided was okay.
But 2066 should not make any difference on performance whatsoever, as it only changes code structure, not behavior, so the playthrough being orders of magnitude worse makes me very, very puzzled. Especially if the generation itself didn't change? Assuming that this benchmark is correct, then it might be a bug on 2066

> Would you like me to leave this on 2206 as well? I did notice it has merge conflicts still
Ah, I didn't notice that 993 caused conflicts.... I'll go fix them tonight.
As for leaving the benchmark info, yes, if 2066 does indeed cause a huge performance hit on playthrough like you seem to be saying, It's probably extremely valuable as a review comment on 2066

**Comment by axe-y** [comment]
I am not safe with using prog_items, but look good.

**Comment by Berserker66** [comment]
> I have no clue why this makes the code faster in any way. But I looked over it, and the code itself seems fine.
> 
> It sure would be nice if it could be checked against the unit tests from #2066 though
> 
> I suspect that the strategy used here might be usable in Stardew as well, so if you could explain it to me a little bit, it might let me make a big improvement there as well

If you look at the implementation of state.has_item this should hopefully become somewhat obvious. We create a pseudo item for the total coins and award / substract that many when a bundle enters or exits state.

**Review by axe-y — APPROVED** [review]

**Review by ThePhar — APPROVED** [review]

**Pull request merged**